### ISSUE TITLE
What algorithm does the "rank:pairwise" used?

### ISSUE THREAD

**Issue by yyyxy** [dmlc/xgboost#5079]
I have checked official documents, which are defined as follows:
rank:pairwise: Use LambdaMART to perform pairwise ranking where the pairwise loss is minimized
rank:ndcg: Use LambdaMART to perform list-wise ranking where Normalized Discounted Cumulative Gain (NDCG) is maximized
rank:map: Use LambdaMART to perform list-wise ranking where Mean Average Precision (MAP) is maximized

Maybe I misunderstood before. I always thought that LambdaMART is a listwise algorithm. How can it do pairwise task and listwise task at the same time?
And "rank:pairwise: Use LambdaMART to perform pairwise ranking where the pairwise loss is minimized"
What is the formula for the loss function defined here?
I was confused and eager for the answer.

**Comment by hcho3** [comment]
LambdaMART operates on a list (query group), and the way the training loss is reduced is by sampling many pairs: https://github.com/dmlc/xgboost/blob/b9dbfe0931dace71ad3e9486dfce6340027b6f63/src/objective/rank_obj.cu#L844-L866

The description of `rank:ndcg` and `rank:map` are quite misleading, since they do almost same thing as `rank:pairwise` and only difference is the loss definition, where the pairwise loss is revised to accommodate a change in NDCG or MAP. In this snippet,
https://github.com/dmlc/xgboost/blob/b9dbfe0931dace71ad3e9486dfce6340027b6f63/src/objective/rank_obj.cu#L867-L868
`GetLambdaWeight()` function does nothing when the objective `rank:pairwise`. For the other objectives (`rank:map`, `rank:ndcg`), `GetLambdaWeight()` returns weight adjustments.

**Comment by trivialfis** [comment]
I think the design is faithful to original algorithm description as that's the strength of lanmdamart, you can put in other metrics as well.
